components/calendar/src/types.rs

```diff
@@ -514,79 +557,108 @@ impl ValidMonthCode {
+    /// This may not necessarily be the canonical month for a month in cases where a month has different
     /// formatting in a leap year, for example Adar/Adar II in the Hebrew calendar in a leap year has
-    /// the standard code M06, but for formatting specifically the Hebrew calendar will return M06L since it is formatted
-    /// differently.
+    /// the standard month `Month::new(6)`, but for formatting specifically the Hebrew calendar will return
+    /// `Month::leap(6)`, since it is formatted differently.
     ///
     /// [`Date::try_new_from_codes`]: crate::Date::try_new_from_codes
     /// [`Date::try_from_fields`]: crate::Date::try_from_fields
-    pub formatting_code: MonthCode,
+    pub formatting: Month,
+
+    /// The [`Month::code()`] of [`Self::standard`].
+    #[deprecated(since = "2.2.0", note = "use `standard.code()")]
+    pub standard_code: MonthCode,
 
-    /// Same as [`Self::formatting_code`] but with invariants validated.
-    pub(crate) valid_formatting_code: ValidMonthCode,
+    /// The [`Month::code()`] of [`Self::formatting`].
+    #[deprecated(since = "2.2.0", note = "use `formatting.code()")]
+    pub formatting_code: MonthCode,
 }
 
 impl MonthInfo {
-    pub(crate) fn non_lunisolar(number: u8) -> Self {
-        Self::for_code_and_ordinal(ValidMonthCode::new_unchecked(number, false), number)
-    }
-
-    pub(crate) fn for_code_and_ordinal(code: ValidMonthCode, ordinal: u8) -> Self {
+    /// Construct a [`MonthInfo`] with the same standard and formatting month
+    pub(crate) fn new_standard<C: crate::calendar_arithmetic::DateFieldsResolver>(
+        c: &C,
+        date: ArithmeticDate<C>,
+    ) -> Self {
+        let ordinal = date.month();
+        let standard = c.month_from_ordinal(date.year(), ordinal);
+        let standard_code = standard.code();
+        #[allow(deprecated)] // field-level allows don't work at 1.83 MSRV
         Self {
             ordinal,
-            standard_code: code.to_month_code(),
-            valid_standard_code: code,
-            formatting_code: code.to_month_code(),
-            valid_formatting_code: code,
+            standard,
+            formatting: standard,
+            #[allow(deprecated)]
+            standard_code,
+            #[allow(deprecated)]
+            formatting_code: standard_code,
         }
     }
 
-    /// Gets the month number. A month number N is not necessarily the Nth month in the year
-    /// if there are leap months in the year, rather it is associated with the Nth month of a "regular"
-    /// year. There may be multiple month Ns in a year
-    pub fn month_number(self) -> u8 {
-        self.valid_standard_code.number()
+    /// Returns the month number of the `standard` [`Month`].
+    ///
+    /// A month number N is not necessarily the Nth month in the year if there are leap
+    /// months in the year, rather it is associated with the Nth month of a "regular"
```

Comment:
Nit: Remove the "rather" part of the sentence. There isn't a general correlation between the N of a leap year and the N of a common year.